src/include/hardware.h

```diff
@@ -16,6 +16,32 @@
         (v) = (val);        \
     } while (0)
 
+/*
+ * WT_READ_ONCE --
+ *
+ * Ensure a single read from memory in the source code produces a single read from memory in the
+ * compiled output.
+ *
+ * The compiler is allowed to optimize loads from memory in multiple ways such as "load fusing",
+ * where the compiler takes multiple reads from a memory location and merges them into a single read
+ * instruction, and "invented loads" where the compiler takes a single load from memory and converts
+ * it into multiple read instructions.
+ *
+ * WiredTiger has many algorithms where threads are allowed to concurrently access and modify the
+ * same memory location, but to do this safely we need to precisely control how reads to memory are
+ * performed. This macro gives us control over this.
+ *
+ * GCC and Clang have a __typeof__ compiler builtin which allows us to temporarily cast the value
+ * being read as a volatile and achieve volatile semantics. For other compilers we'll fall back on
+ * inserted a read barrier after the read (our pre-existing implementation) which prevents invented
+ * and fused loads for this variable in the code following the expression.
+ */
+#if defined(__GNUC__) || defined(__clang__)
+#define WT_READ_ONCE(v, val) (v) = (*(volatile __typeof__(val) *)&(val));
+#else
+#define WT_READ_ONCE(v, val) WT_ORDERED_READ(v, val)
```

Comment:
I was thinking about this but don't see it as a natural path forward without causing perf issues on windows. For example WT-11414 wants to replace volatile generations with read_once. But if we did that everywhere the code would slow down on windows? So we couldn't introduce it there. There's also another PR #9456 that did have a volatile which could've used READ_ONCE, it went with locking instead but I don't know how comfortable I'd be introducing a tonne of read barriers on Windows.

Reply:
As a result of being unable to implement READ_ONCE on windows due to some MSVC issues we're going ahead with a read barrier.